### PR TITLE
POST/PATCH/JSON PATCH implementation

### PR DESCRIPTION
**Pull request by joshuawscott**
This implements the POST/PATCH/JSON PATCH methods.

I also updated the Trafficcontroller example so that it shows how to add an IP to a trafficcontroller pool. The example doesn't work for an empty TC pool, though, but I was hesitant to make it more complex when it's just a demo.

### CODE REVIEW

File: examples/zone/main.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		Password: *passPtr,
+	})
+
+	var url string
+
+	if *zonePtr == "" {
+		url = "/zones/"
+	} else {
+		url = "/zones/" + *zonePtr
+	}
```
Comment (nathanmonteleone) [inline examples/zone/main.go:34]: Does go get mad if you concatenate `""`?  If not then can we just do `url = "/zones/" + *zonePtr` without the conditional?

File: examples/zone/main.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+	// Make a GET request to the /status endpoint
+	resp, err := apiConn.Get(url)
+	if err != nil {
+		fmt.Print("Error in apiConn.Get: ")
+		fmt.Println(err)
+		return
+	}
+	// The body of the response is an http.Response.Body ReadCloser. It needs to be closed after being read.
+	defer resp.Body.Close()
```
Comment (nathanmonteleone) [inline examples/zone/main.go:44]: this is cool.

File: examples/zone/main.go (hunk 1 above)
Comment (nathanmonteleone) [reply examples/zone/main.go:34]: oh hell, this is just an example.  Never mind unless you're in a code-golfing mood.

File: pkg/ultradns/ultradns.go [hunk 3]
```diff
@@ -110,3 +110,80 @@ func (apiConn *APIConnection) Post(url string, body io.Reader) (resp *http.Respo
...
 }
+
+// Put executes a PUT request at the given url using the APIConnection's client and credentials.
+// This function imitates the http.Post API, but does not require a Content-Type as the type is always set to
+// 'application/json'
+//
+// error will be non-nil when:
+// * encountering an error authorizing
+// * Failing to connect to the API server
+// * When getting an HTTP status code of >= 400
```
Comment (nathanmonteleone) [inline pkg/ultradns/ultradns.go:121]: excellent docs

File: pkg/ultradns/ultradns.go [hunk 4]
```diff
@@ -110,3 +110,80 @@ func (apiConn *APIConnection) Post(url string, body io.Reader) (resp *http.Respo
 	}
 	return resp, err
 }
+
+// Put executes a PUT request at the given url using the APIConnection's client and credentials.
+// This function imitates the http.Post API, but does not require a Content-Type as the type is always set to
```
Comment (nathanmonteleone) [inline pkg/ultradns/ultradns.go:115]:
```suggestion
// This function imitates the http.Put API, but does not require a Content-Type as the type is always set to
```

Comment (nathanmonteleone) [reply pkg/ultradns/ultradns.go:115]: (I don't know if "put" or "post" is correct here, but it looked like copypasta?)

File: pkg/ultradns/ultradns.go [hunk 5]
```diff
@@ -110,3 +110,80 @@ func (apiConn *APIConnection) Post(url string, body io.Reader) (resp *http.Respo
...
+	req.Header.Add("Content-Type", "application/json")
+	resp, err = apiConn.Client.Do(req)
+	if err == nil {
+		err = ultradns.GetError(resp)
+	}
+	return resp, err
+}
+
+// Patch executes a PATCH request at the given url using the APIConnection's client and credentials.
+// This function imitates the http.Post API, but does not require a Content-Type as the type is always set to
```
Comment (nathanmonteleone) [inline pkg/ultradns/ultradns.go:140]: should this read http.Patch API?

File: pkg/ultradns/ultradns.go [hunk 6]
```diff
@@ -110,3 +110,80 @@ func (apiConn *APIConnection) Post(url string, body io.Reader) (resp *http.Respo
...
+	if err == nil {
+		err = ultradns.GetError(resp)
+	}
+	return resp, err
+}
+
+// JSONPatch executes a PATCH request at the given url using the APIConnection's client and credentials.
+// JSON Patch is a special form of PATCH request that UltraDNS provides to allow partialy altering
+// a record set. See the UltraDNS REST API for more detail.
+// This function imitates the http.Post API, but does not require a Content-Type as the type is always set to
```
Comment (nathanmonteleone) [inline pkg/ultradns/ultradns.go:167]: http.Patch API?

File: examples/zone/main.go (hunk 1 above)
Comment (joshuawscott) [reply examples/zone/main.go:34]: Good call

File: pkg/ultradns/ultradns.go (hunk 4 above)
Comment (joshuawscott) [reply pkg/ultradns/ultradns.go:115]: So, the http library only has Get() Post() and Head() I believe - therefore this should be `Post`
https://golang.org/pkg/net/http/

File: pkg/ultradns/ultradns.go (hunk 5 above)
Comment (joshuawscott) [reply pkg/ultradns/ultradns.go:140]: Same as above - there's only an http.Post API.

File: examples/zone/main.go (hunk 1 above)
Comment (joshuawscott) [reply examples/zone/main.go:34]: fixed

**Review by cjonesy — APPROVED** [review]
I don't know UltraDns very well, but the code itself makes sense.  I particularly like the example apps. :shipit:

File: pkg/ultradns/ultradns.go [hunk 7]
```diff
@@ -110,3 +110,80 @@ func (apiConn *APIConnection) Post(url string, body io.Reader) (resp *http.Respo
...
+		return nil, err
+	}
+	req.Header.Add("Authorization", "Bearer "+apiConn.Authorization.AccessToken)
+	req.Header.Add("Content-Type", "application/json-patch+json")
+	resp, err = apiConn.Client.Do(req)
+	if err == nil {
+		err = ultradns.GetError(resp)
+	}
+	return resp, err
+}
```
Comment (nathanmonteleone) [inline pkg/ultradns/ultradns.go:189]: I'm on the fence as to whether lines 182-188 should be a helper function (they're repeated in most of these new fn's except for the content type, as far as I can tell).

Comment (joshuawscott) [reply pkg/ultradns/ultradns.go:189]: Yeah, I'm with you there. I can make another PR to collapse it down if possible.

**Review by nathanmonteleone — APPROVED** [review]
:shipit:

**Pull request merged**